doc/manpages/distcheck.1
.TH distcheck 1 "December 3, 2010" "Version 2.9" "USER COMMANDS"
.SH NAME
distcheck \- check installability of packages according to metadata

.SH SYNOPSIS
\fBdistcheck\fR [option] ... \fIuri\fR
.br
\fBdebcheck\fR [option] ... \fIfile\fR
.br
\fBrpmcheck\fR [option] ... \fIfile\fR
.br
\fBeclipsecheck\fR [option] ... \fIfile\fR

.SH DESCRIPTION
.B distcheck
determines, for a set of package control stanzas, called the
repository, whether packages of the repository can be installed
relative to the repository according to the inter-package
relationsships expressed in the package control stanzas. The exact set
of relevant control fields and their meaning depends on the type of the
repository.
.P
The constraint solving algorithm is complete, that is it finds a
solution whenever there exists one, even for multiple disjunctive
dependencies and deep package conflicts. This problem is
computationally infeasible in theory (that is, NP-complete), but can
be solved very efficiently for package repositories that actually occur
in practice.

.SH INPUT SPECIFICATION
Currently supported repository types are debian, rpm, and eclipse. The
.B distcheck
tool expects its input specifiation in the form
.I type://pathname
where
.I type
is one of \fBdeb\fR, \fBsynthesis\fR, \fBhdlist\fR or \fBeclipse\fR, and
.I pathname
is the pathname of a file containing the input. The package metadata
found in that file must correspond to the
.I type
given in the URI.

When invoked as \fItype\fBcheck\fR, the type of the repository is
assumed to be \fItype\fR, and the input is simply given in form of a
pathname of a file containing the repository.
.B debcheck
accepts in addition 
.B --
as input specification, with the effect that input is taken from
standard input.	
  
.SH INPUT FORMATS

.SS Debian
The input file has to contain stanzas in the format of deb-control(5),
separated by one blank line. For instance, the Packages files as found
on a Debian mirror server, or in the directory
\fI/var/lib/apt/lists/\fR of a Debian system, are suitable as input to
\fBdebcheck\fR.
.P
Installability of packages is analyzed according to their
\fBDepends\fR, \fBConflicts\fR, and \fBProvides\fR fields with their
meaning as of Debian policy version 3.9.0. \fBPre-depends\fR are
treated like \fBDepends\fR, and \fBBreaks\fR are treated like
\fBConflicts\fR.
.P
In the case of Debian, it is not possible to install at the same time
two packages with the same name but different versions.

.SS RPM
The input file can be either a \fIsynthesis\fR file or a \fIhdlist\fR
file.  By default rpmcheck expects a synthesis file as input. To
specify a hdlist file distcheck must be invoked with a file argument
of the form hdlist://

.SS Eclipse
The input is an 822 file containing the encoding of OSGi plugins 
content.xml files.

.SH OPTIONS
.TP
.B \-f \-\-failures
Only show broken packages that fail the installability check.
.TP
.B \-v \-\-successes
Only show packages that do not fail the installability check.
.TP
.B \-e \-\-explain
Explain the results in more detail.
.TP
.B \-\-checkonly
Select a subset of packages to check. By default all packages are checked.
.TP
.B \-v
Enable info / warnings / debug messages. This option may be repeated
up to three times in order to increase verbosity.
.TP
.B \-\-progress
Enable progress bars.
.TP
.B \-h, \-\-help
Display this list of options.

.SH EXIT CODE
The exit code is 0 when all packages that were asked to be checked
(either explicitly with the \-\-checkonly option, or because they
exist in the repository) are available and found to be installable, it
is 1 when a least one of these packages is available but not
installable, and it is 2 when at least one of the packages is not
available but all available packages are found to be
installable. Hence, the exit code 2 can occur only when the
\-\-checkonly option is used.

.SH EXAMPLE
Check which packages in a particular distribution are not installable and why:

  debcheck \-\-failures \-\-explain < Packages

where Packages is the file pertaining to that distribution, as for instance
found in the directory \fB/var/lib/apt/lists\fR.

.SH NOTE
distcheck is a complete reimplementation of edos-debcheck, written for the
EDOS project.

.SH AUTHOR
The first version of debcheck had been written by Jérôme
Vouillon for the EDOS project. The current version has been rewritten
on the basis of the dose3 library by Pietro Abate.

.SH SEE ALSO
.BR deb-control (5)

<http://www.edos-project.org> is the home page of the EDOS project.
<http://www.mancoosi.org> is the home page of the Mancoosi project.



applications/distcheck.ml
(**************************************************************************************)
(*  Copyright (C) 2009 Pietro Abate <pietro.abate@pps.jussieu.fr>                     *)
(*  Copyright (C) 2009 Mancoosi Project                                               *)
(*                                                                                    *)
(*  This library is free software: you can redistribute it and/or modify              *)
(*  it under the terms of the GNU Lesser General Public License as                    *)
(*  published by the Free Software Foundation, either version 3 of the                *)
(*  License, or (at your option) any later version.  A special linking                *)
(*  exception to the GNU Lesser General Public License applies to this                *)
(*  library, see the COPYING file for more information.                               *)
(**************************************************************************************)

open ExtLib
open Debian
open Common
open Diagnostic

module Options = struct
  open OptParse
  let description = "Report the broken packages in a package list"
  let options = OptParser.make ~description
  include Boilerplate.MakeOptions(struct let options = options end)

  let successes = StdOpt.store_true ()
  let failures = StdOpt.store_true ()
  let explain = StdOpt.store_true ()
  let uuid = StdOpt.store_true ()
  let checkonly = Boilerplate.vpkglist_option ()
  let architecture = StdOpt.str_option ()
  let distribution = StdOpt.str_option ()
  let release = StdOpt.str_option ()
  let suite = StdOpt.str_option ()
  let summary = StdOpt.store_true ()
  let outfile = StdOpt.str_option ()

  open OptParser
  add options ~short_name:'e' ~long_name:"explain" ~help:"Explain the results" explain;
  add options ~short_name:'f' ~long_name:"failures" ~help:"Only show failures" failures;
  add options ~short_name:'s' ~long_name:"successes" ~help:"Only show successes" successes;

  add options ~long_name:"checkonly" ~help:"Check only these package" checkonly;
  add options ~long_name:"summary" ~help:"Print a detailed summary" summary;

  add options ~long_name:"distrib" ~help:"Set the distribution" distribution;
  add options ~long_name:"release" ~help:"Set the release name" release;
  add options ~long_name:"suite" ~help:"Set the release name" suite;
  add options ~long_name:"arch" ~help:"Set the default architecture" architecture;
  add options ~short_name:'u' ~long_name:"uid" ~help:"Generate a unique identifier for the output document" uuid;

  add options ~short_name:'o' ~long_name:"outfile" ~help:"output file" outfile;
end

let debug fmt = Util.make_debug "Distcheck" fmt
let info fmt = Util.make_info "Distcheck" fmt
let warning fmt = Util.make_warning "Distcheck" fmt
let fatal fmt = Util.make_fatal "Distcheck" fmt

let timer = Util.Timer.create "Solver" 

let main () =
  let posargs =
    let args = OptParse.OptParser.parse_argv Options.options in
    match Filename.basename(Sys.argv.(0)),args with
    |("debcheck"|"edos-debcheck"),[] -> ["deb://-"]
    |("debcheck"|"edos-debcheck"),l -> List.map ((^) "deb://") l
    |"eclipsecheck",l -> List.map ((^) "eclipse://") l
    |("rpmcheck"|"edos-rpmcheck"),l -> List.map ((^) "synthesis://") l
    |_,_ -> args
  in
  Boilerplate.enable_debug (OptParse.Opt.get Options.verbose);
  Boilerplate.enable_timers (OptParse.Opt.get Options.timers) ["Solver"];
  let default_arch = OptParse.Opt.opt Options.architecture in
  let (universe,from_cudf,to_cudf) = Boilerplate.load_universe ~default_arch posargs in
  let universe_size = Cudf.universe_size universe in
  let pkglist = 
    if OptParse.Opt.is_set Options.checkonly then 
        List.flatten (
          List.map (function 
            |(p,None) -> Cudf.lookup_packages universe p
            |(p,Some(c,v)) ->
                let filter = Some(c,snd(to_cudf (p,v))) in
                Cudf.lookup_packages ~filter universe p
          ) (OptParse.Opt.get Options.checkonly)
        )
    else []
  in
  let pp pkg =
    let (p,v) = from_cudf (pkg.Cudf.package,pkg.Cudf.version) in 
    let l = 
      List.filter_map (fun k ->
        try Some(k,Cudf.lookup_package_property pkg k)
        with Not_found -> None
      ) ["architecture";"source";"sourceversion"]
    in (p,v,l)
  in
  info "Solving..." ;
  let failure = OptParse.Opt.get Options.failures in
  let success = OptParse.Opt.get Options.successes in
  let explain = OptParse.Opt.get Options.explain in
  let summary = OptParse.Opt.get Options.summary in
  let fmt =
    if OptParse.Opt.is_set Options.outfile then
      let oc = open_out (OptParse.Opt.get Options.outfile) in
      Format.formatter_of_out_channel oc
    else
      Format.std_formatter
  in
  let results = Diagnostic.default_result universe_size in

  if OptParse.Opt.get Options.uuid then
    Format.fprintf fmt "uid: %s@." (Util.uuid ());
  if OptParse.Opt.is_set Options.distribution then
    Format.fprintf fmt "distribution: %s@." (OptParse.Opt.get Options.distribution);
  if OptParse.Opt.is_set Options.release then
    Format.fprintf fmt "release: %s@." (OptParse.Opt.get Options.release);
  if OptParse.Opt.is_set Options.suite then
    Format.fprintf fmt "suite: %s@." (OptParse.Opt.get Options.suite);
  if OptParse.Opt.is_set Options.architecture then
    Format.fprintf fmt "architecture: %s@." (OptParse.Opt.get Options.architecture);

  if failure || success then Format.fprintf fmt "@[<v 1>report:@,";
  let callback d =
    if summary then Diagnostic.collect results d ;
    Diagnostic.fprintf ~pp ~failure ~success ~explain fmt d
  in
  Util.Timer.start timer;
  let i =
    if OptParse.Opt.is_set Options.checkonly then 
      Depsolver.listcheck ~callback universe pkglist
    else
      Depsolver.univcheck ~callback universe 
  in
  ignore(Util.Timer.stop timer ());

  if failure || success then Format.fprintf fmt "@]@.";
 
  let nb = universe_size in
  let nf = List.length pkglist in
  Format.fprintf fmt "background-packages: %d@." nb;
  Format.fprintf fmt "foreground-packages: %d@." (if nf = 0 then nb else nf);
  Format.fprintf fmt "broken-packages: %d@." i;
 
  if summary then 
    Format.fprintf fmt "@[%a@]@." (Diagnostic.pp_summary ~pp ()) results;
;;

main () ;;
